spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryJsonDataWriter.java

```diff
@@ -0,0 +1,156 @@
+import org.json.JSONArray;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Helper class for using BigQuery storage write API in exactly once delivery mode */
+public class BigQueryJsonDataWriter implements AutoCloseable {
+
+  private final JsonStreamWriter streamWriter;
+  // Track the number of in-flight requests to wait for all responses before shutting down.
+  private final Phaser inflightRequestCount = new Phaser(1);
+
+  private final Object lock = new Object();
+
+  private final Logger logger = LoggerFactory.getLogger(BigQueryJsonDataWriter.class);
+
+  @GuardedBy("lock")
+  private RuntimeException error = null;
+
+  /**
+   * @param parentTable against which the writer has to be initialized
+   * @param client BigQueryWriteClient reference which has to be used for writing to the database
+   */
+  public BigQueryJsonDataWriter(TableName parentTable, BigQueryWriteClient client)
+      throws DescriptorValidationException, IOException, InterruptedException {
+    // Initialize a write stream for the specified table.
+    // For more information on WriteStream.Type, see:
+    // https://googleapis.dev/java/google-cloud-bigquerystorage/latest/com/google/cloud/bigquery/storage/v1/WriteStream.Type.html
+    WriteStream stream = WriteStream.newBuilder().setType(WriteStream.Type.PENDING).build();
+
+    CreateWriteStreamRequest createWriteStreamRequest =
+        CreateWriteStreamRequest.newBuilder()
+            .setParent(parentTable.toString())
+            .setWriteStream(stream)
+            .build();
+    WriteStream writeStream = client.createWriteStream(createWriteStreamRequest);
+
+    // Use the JSON stream writer to send records in JSON format.
+    // For more information about JsonStreamWriter, see:
+    // https://googleapis.dev/java/google-cloud-bigquerystorage/latest/com/google/cloud/bigquery/storage/v1beta2/JsonStreamWriter.html
+    streamWriter =
+        JsonStreamWriter.newBuilder(writeStream.getName(), writeStream.getTableSchema()).build();
+  }
+
+  /**
+   * @param data JSONArray to be appended
+   * @param offset offset at which data has to be added
+   */
+  public void append(JSONArray data, long offset)
+      throws DescriptorValidationException, IOException, ExecutionException {
+    synchronized (this.lock) {
+      // If earlier appends have failed, we need to reset before continuing.
+      if (this.error != null) {
+        throw this.error;
+      }
+    }
+    // Append asynchronously for increased throughput.
+    ApiFuture<AppendRowsResponse> future = streamWriter.append(data, offset);
+    ApiFutures.addCallback(future, new AppendCompleteCallback(), MoreExecutors.directExecutor());
+    // Increase the count of in-flight requests.
+    inflightRequestCount.register();
+  }
+
+  /**
+   * Call this method before committing the stream
+   *
+   * @param client BigQueryWriteClient
+   */
+  public void finalizeWriteStream(BigQueryWriteClient client) {
```

Comment:
Since the client got passed in the constructor, would it make sense to persist it there as a field, and have `finalizeWriteStream()` method be no-argument?

Reply:
sure. updated the code